File: leetcode/src/java/LinkedListCycle.java
import java.util.HashSet;
import java.util.Set;

public class LinkedListCycle extends BaseListNode {
    public static void main(String[] args) {
        LinkedListCycle instance = new LinkedListCycle();
        System.out.println(instance.hasCycle(generateListNode(new int[]{3, 2, 0, -4})));
    }
    public boolean hasCycle(ListNode head) {
        boolean result = false;
        Set<ListNode> set = new HashSet<ListNode>();
        if (head != null) {
            set.add(head);
            ListNode temp = head;
            while (temp.next != null) {
                if (set.contains(temp.next)) {
                    result = true;
                    break;
                } else {
                    set.add(temp.next);
                    temp = temp.next;
                }
            }
        }
        return result;
    }
}

File: leetcode/src/python/linked_list_cycle.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class LinkedListCycle:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        node_set = set()
        if head is not None:
            current = head
            while current is not None:
                if current in node_set:
                    return True
                else:
                    node_set.add(current)
                current = current.next
        return False